executor/builder.go

```diff
@@ -656,12 +660,74 @@ func (b *executorBuilder) buildRevoke(revoke *ast.RevokeStmt) Executor {
 }
 
 func (b *executorBuilder) buildDDL(v *plannercore.DDL) Executor {
-	e := &DDLExec{
+	if b.ctx.GetSessionVars().CreateTableInsertingID != 0 {
+		// in a 'inserting data from select' state of creating table.
+		return b.buildTableInserter(v, b.ctx.GetSessionVars().CreateTableInsertingID)
+	}
+	return &DDLExec{
 		baseExecutor: newBaseExecutor(b.ctx, v.Schema(), v.ExplainID()),
 		stmt:         v.Statement,
 		is:           b.is,
 	}
-	return e
+}
+
+// buildTableInserter builds a CreateTableInsertExec to insert data when creating table by 'create table ... select'
+func (b *executorBuilder) buildTableInserter(v *plannercore.DDL, tableID int64) Executor {
+	stmt, ok := v.Statement.(*ast.CreateTableStmt)
+	if !ok || v.InsertPlan.SelectPlan == nil {
+		b.err = errors.Errorf("Unexpected plan: %s", v.Statement.Text())
+		return nil
+	}
+	selectExec := b.build(v.InsertPlan.SelectPlan)
+
+	m := meta.NewMeta(b.ctx.Txn(true))
+	dbInfo, ok := b.is.SchemaByName(stmt.Table.Schema)
+	if !ok {
+		b.err = infoschema.ErrDatabaseNotExists.GenWithStackByArgs(stmt.Table.Schema.L)
+		return nil
+	}
+	tbInfo, err := m.GetTable(dbInfo.ID, tableID)
+	if err != nil {
+		b.err = errors.Trace(err)
+		return nil
+	}
+
+	// We have to create an allocator here, which is not the one created by InfoSchema, but it should be fine.
+	alloc := autoid.NewAllocator(b.ctx.GetStore(), tbInfo.GetDBID(dbInfo.ID))
+	tbl, err := tables.TableFromMeta(alloc, tbInfo)
+	if err != nil {
+		b.err = errors.Trace(err)
+		return nil
+	}
+	// The operation of the minus 1 to make sure that the current value doesn't be used,
+	// the next Alloc operation will get this value.
+	// Its behavior is consistent with MySQL.
+	if err = tbl.RebaseAutoID(nil, tbInfo.AutoIncID-1, false); err != nil {
```

Comment:
The auto increment id will be rebased after the table create finished, create select rebase it here, so if the create_select is true, we should not set auto-id in https://github.com/pingcap/tidb/blob/master/ddl/ddl_api.go#L964?

Reply:
Could we add a test case to test `create table auto_increment= xxx select `?

Reply:
1. You're right, I didn't treat here carefully, the current codes just work because `Allocator`.`Rebase` prevented base id rolling back.
2. There is a test case already, but now I think it can be enhanced.

Thanks for the review